src/relay/pass/to_anf.cc

```diff
@@ -0,0 +1,418 @@
+ * But what is the lowest common ancestor?
+ * The problem is here because you also need to lift lambda.
+ * And lambda bind variable.
+ * So while you are lifting stuff,
+ * the scope that dictate how you should lift stuff is also being lifted too.
+ *
+ * The insight to the algorithm is that, we gotta think backward.
+ * Instead of the usual recursion, where parent scope are built from children scope,
+ * Children scope should be built from the parent scope.
+ *
+ * The scope of the whole expr is global.
+ * The scope of any subexpr, is the lowest common ancestor of all it's dependent,
+ * so they can all use it.
+ *
+ * In order to do this, we need to:
+ * 0: find dependency between expr.
+ * 1: toposort it.
+ * 2: for all new scope, create a LetList ll.
+ * 3: flush every value to the ll of the lowest common ancestor.
+ * 4: collect the values.
+ */
+Expr ToANF(const Expr& e, const Module& m, std::set<GlobalVar>* gv);
+
+struct ScopeNode;
+using Scope = std::shared_ptr<ScopeNode>;
+using OptionalScope = Scope;
+
+/* Invariant: when parent is null level is 0
+ *
+ * Invariant: when parent is notempty level is 1 + parent->level
+ */
+struct ScopeNode {
+  size_t level;
+  Scope parent;
+  std::shared_ptr<LetList> ll = std::make_shared<LetList>();
+  explicit ScopeNode(const Scope& parent) : level(1 + parent->level), parent(parent) { }
+  ScopeNode() : level(0) { }
+};
+
+Scope ChildScope(const Scope& s) {
+  return std::make_shared<ScopeNode>(s);
+}
+
+OptionalScope LCA(OptionalScope lhs, OptionalScope rhs) {
+  if (!lhs) {
+    return rhs;
+  } else if (!rhs) {
+    return lhs;
+  } else {
+    while (lhs != rhs) {
+      if (lhs->level > rhs->level) {
+        lhs = lhs->parent;
+      } else if (lhs->level < rhs->level) {
+        rhs = rhs->parent;
+      } else {
+        lhs = lhs->parent;
+        rhs = rhs->parent;
+      }
+    }
+    return lhs;
+  }
+}
+
+struct ExprPair {
+  Expr lhs;
+  Expr rhs;
+  bool operator==(const ExprPair& oth) const {
+    return lhs == oth.lhs && rhs == oth.rhs;
+  }
+  ExprPair(const Expr& lhs, const Expr& rhs) : lhs(lhs), rhs(rhs) { }
+};
+
+struct ExprPairHash {
+  size_t operator()(const ExprPair& p) const {
+    return dmlc::HashCombine(NodeHash()(p.lhs), NodeHash()(p.rhs));
+  }
+};
+
+using Graph = std::unordered_map<Expr,
+                                 std::unordered_set<Expr, NodeHash, NodeEqual>,
+                                 NodeHash,
+                                 NodeEqual>;
+
+/* Invariant: after visited dependency of such expr will be filled.
+ *
+ * Invariant: dependency and dependent is inverse of each other.
+ */
+struct ExprDep : ExprFunctor<void(const Expr&)> {
+  void VisitExpr(const Expr& e) final {
+    if (dependency.count(e) == 0) {
+      dependency.insert({e, {}});
+      ExprFunctor<void(const Expr&)>::VisitExpr(e);
+    }
+  }
+
+  void Depend(const Expr& parent, const Expr& child) {
+    CHECK_NE(dependency.count(parent), 0);
```

Comment:
Is it possible that a node is visited multiple twice and then this assertion fails, since you are using `ExprFunctor<void(const Expr&)>`)?